### ISSUE TITLE
Windows testing errors for 1.10.2

### ISSUE THREAD

**Issue by charris** [numpy/numpy#6807]
Christoph reports on 1.10.x

On Python 2.x, 32 and 64 bit:
```
==================================================
FAIL: test_mvoid_multidim_print (test_core.TestMaskedArray)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python27-x64\lib\site-packages\numpy\ma\tests\test_core.py", line 716, in test_mvoid_multidim_print
    assert str(t_ma[0]) == "([1, --, 3],)"
AssertionError
```
On Python 2.x, 32 bit:
```
===================================================
FAIL: test_umath.test_rint_big_int
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python27\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "X:\Python27\lib\site-packages\numpy\core\tests\test_umath.py", line 1939, in test_rint_big_int
    assert_equal(val, np.rint(val))
  File "X:\Python27\lib\site-packages\numpy\testing\utils.py", line 355, in assert_equal
    raise AssertionError(msg)
AssertionError:
Items are not equal:
 ACTUAL: 4607998452777363968L
 DESIRED: 4.6079984527773635e+18
```
The second looks like a promotion of `long` to `float`, not sure about the first. Can someone running windows track these down? I don't think they are serious, but it would be good to know where they come from.

Note that in the first test `assert_` is not used.

**Comment by njsmith** [comment]
Are these new in 1.10.2rc versus 1.10.1, or new in 1.10.2rc versus 1.9,
or...?

**Comment by charris** [comment]
The first is from 2015-7-19 (1.10.0) and the second from 2015-11-17 (1.10.2 backport). The second looks like a test problem to me, not sure about the first.

**Comment by njsmith** [comment]
Second one is probably #6685?

**Comment by njsmith** [comment]
@cgohlke: it'd probably be helpful also for the first error if you could check what `str(t_ma[0])` is returning.

**Comment by charris** [comment]
Actually, I'm wondering if the second test is a library problem for 32 git windows.

**Comment by charris** [comment]
`rint` looks new in VS 2013, so this may be a problem with our fallback implementation, although I don't see anything that should be a problem unless the system `floor` function also has a problem.

**Comment by matthew-brett** [comment]
I seem to remember that the rint test is failing on Python 2.7, therefore a problem on MSVC 2008?

We found this via some precision checks in `nibabel`.  Logs at the time - e.g. https://ci.appveyor.com/project/Eric89GXL/nibabel/build/1.0.362 - show failures on Python 2.7, but not Python 3.4. 

I believe that this error appears for numpy 1.10 (not 1.9.x) for both conda and Christoph's wheels - so I guess it's some interaction between MSVC and numpy that has been exposed in 1.10.

For example here's an earlier appveyor log that is running the same tests and passing, with numpy 1.9.3 : https://ci.appveyor.com/project/Eric89GXL/nibabel/build/1.0.263/job/28gjux771njwpu4h

**Comment by charris** [comment]
The invalid values in rint for 1.9 are not related? They are the only thing that seems common between the 1.9 and 1.10.1 runs on 32 bits. The only error is the round trip test.

**Comment by matthew-brett** [comment]
I hadn't noticed those `invalid value` warnings.  I don't know if they are relevant - I just looked at the other builds on 32-bit Linux and other platforms - we don't seem to have those warnings:

http://nipy.bic.berkeley.edu/builders/nibabel-py2.6-32/builds/338/steps/shell_5/logs/stdio

Still - the test that's failing here definitely passes with MSVC numpy 1.9.3 on Windows on my machines, and fails for 1.10.1.

**Comment by charris** [comment]
That test isnt in 1.9...

**Comment by matthew-brett** [comment]
No, sorry, I meant that, if you do the same calculation in numpy 1.9, you get the right answer:

```
PS C:\repos> c:/python27/scripts/virtualenv venv271
Ignoring indexes: https://pypi.python.org/simple
Collecting pip
Collecting wheel
Collecting setuptools
Installing collected packages: pip, wheel, setuptools
Successfully installed pip-7.1.2 setuptools-18.5 wheel-0.26.0
PS C:\repos> venv271\Scripts\activate
(venv271) PS C:\repos> pip install C:\Users\mb312\Downloads\numpy-1.9.2+mkl-cp27-none-win32.whl
Processing c:\users\mb312\downloads\numpy-1.9.2+mkl-cp27-none-win32.whl
Installing collected packages: numpy
Successfully installed numpy-1.9.2
(venv271) PS C:\repos> python
Python 2.7.8 (default, Jun 30 2014, 16:03:49) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> ds = 4607998452777363968
>>> int(float(ds))
4607998452777363968L
>>> int(np.rint(ds))
4607998452777363968L
>>>
```

**Comment by cgohlke** [comment]
>  it'd probably be helpful also for the first error if you could check what str(t_ma[0]) is returning.

```
======================================================================
FAIL: test_mvoid_multidim_print (test_core.TestMaskedArray)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python27-x64\lib\site-packages\numpy\ma\tests\test_core.py", line 716, in test_mvoid_multidim_print
    assert str(t_ma[0]) == "([1, --, 3],)", t_ma[0]
AssertionError: ([1L, --, 3L],)
```

**Comment by cgohlke** [comment]
Re `rint` on win32-py2.x: Numpy 1.10.x uses SSE2. This could be another msvc9 compiler bug. In fact disabling optimizations for the `npy_rint` function fixes the test failure (do not apply! this is not a final fix):
```diff
diff --git a/numpy/core/src/npymath/npy_math.c.src b/numpy/core/src/npymath/npy_math.c.src
index 7f62810..fe1f0ae 100644
--- a/numpy/core/src/npymath/npy_math.c.src
+++ b/numpy/core/src/npymath/npy_math.c.src
@@ -260,6 +260,7 @@ double npy_atanh(double x)
 #endif

 #ifndef HAVE_RINT
+#pragma optimize("", off)
 double npy_rint(double x)
 {
     double y, r;
@@ -280,6 +281,7 @@ double npy_rint(double x)
     }
     return y;
 }
+#pragma optimize("", on)
 #endif

 #ifndef HAVE_TRUNC
```

**Comment by njsmith** [comment]
> another msvc9 compiler bug

Goddammit. Why do we even still support msvc? icc builds are widely available, and at least one of (icc, mkl) is currently mandatory on windows if you want reasonable performance anyway. Is our support for msvc accomplishing anything beyond giving people (companies) an excuse to keep freeloading instead of helping fix the problem properly by funding mingw-w64?

(No I don't really mean this. Maybe.)

**Comment by njsmith** [comment]
> `AssertionError: ([1L, --, 3L],)`

OK, so this means the masked array thing is just a harmless test bug, not a real bug.

**Comment by charris** [comment]
@njsmith What we need is a windows version of 2.7 compiled with a newer compiler. I'm beginning to understand why python 3 uptake is faster on windows...

**Comment by njsmith** [comment]
> What we need is a windows version of 2.7 compiled with a newer compiler

This gets proposed occasionally on python-dev, but it will never actually happen; there's a few different ways they could try to do it but they all create huge messes.

**Comment by charris** [comment]
The errors are fixed by #6817, #6819, and #6813.
